osquery/tables/smart/smart_drives.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <smartmontools/libsmartctl.h>
+#include <smartmontools/smartctl_errs.h>
+
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+#include <osquery/tables.h>
+
+#include "osquery/events/linux/udev.h"
+
+namespace osquery {
+namespace tables {
+
+/* @brief Represents a hardware driver type that SMART api can you use to query
+ * device information.
+ *
+ * @param driver name of SMART controller driver
+ * @param maxID max ID number of which disks on the controller is monitored
+ */
+struct hardwareDriver {
+  std::string driver;
+  size_t maxID;
+};
+
+std::ostream& operator<<(std::ostream& stream,
+                         const std::vector<hardwareDriver>& devs) {
+  for (const auto& dev : devs) {
+    stream << " " << dev.driver;
+  }
+
+  return stream;
+}
+
+/// Static map of supported controller types.
+static const std::map<std::string, hardwareDriver>
+    kSMARTExplicitDriverToDevice = {
+        {"megaraid_sas", hardwareDriver{"megaraid,", 127}},
+        {"hpsa", hardwareDriver{"cciss,", 14}},
+};
+
+/// Gets all devices of class 'Mass storage controller'.
+std::vector<std::string> getStorageCtlerClassDrivers() {
+  std::vector<std::string> results;
+
+  auto devices = SQL::selectAllFrom("pci_devices");
+  for (const auto& device : devices) {
+    if (device.find("pci_class") != device.end() &&
+        device.at("pci_class") == "Mass storage controller" &&
+        device.find("driver") != device.end()) {
+      auto i =
+          std::lower_bound(results.begin(), results.end(), device.at("driver"));
+      if (i == results.end() || device.at("driver") < *i) {
+        results.insert(i, std::move(device.at("driver")));
+      }
+    }
+  }
+
+  return results;
+}
+
+/// Gets supported hardwareDriver from system storage controllers.
+static inline void getSmartCtlDeviceType(
+    const std::vector<std::string>& storageDrivers,
+    std::vector<hardwareDriver>& types) {
+  for (auto const& driver : storageDrivers) {
+    if (kSMARTExplicitDriverToDevice.find(driver) !=
+        kSMARTExplicitDriverToDevice.end()) {
+      types.push_back(kSMARTExplicitDriverToDevice.at(driver));
+    }
+  }
+}
+
+/// Utility function for traversing system devices.
+void walkDevices(std::function<bool(libsmartctl::Client& c,
```

Comment:
This seems a bit unnecessarily convoluted to me, handing a lambda as a parameter to a function. Perhaps you can find a way to restructure this so that either you're using the lambda explicitly below, and don't need to pass it as a parameter here, or you could rewrite some of this to be two explicit functions?